cibyl/sources/zuuld/git_api.py

```diff
@@ -0,0 +1,216 @@
+                    if 'job' in i.keys():
+                        name = i['job']['name']
+                        t.add_job(Job(name=name))
+        return AttributeDictValue("jobs", attr_type=Tenant,
+                                  value={"openstack": t})
+
+
+class ZuulLocal(GitSource):
+    def __init__(self, kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        self.repos = kwargs.get("repos")
+        super(ZuulLocal, self).__init__(repos=self.repos)
+
+    def _validate(self):
+        """
+        Validate that repo is present, or pull the latest changes
+        Locate zuul job source.
+        """
+        # Check repo exists, if yes pull latest changes
+        self.setup()
+
+        # First check for directory then for file
+        for repo in self.repos:
+            if os.path.isdir(os.path.join(repo.get('dest'),
+                                          ZUUL_JOBS_LOCATION[0])):
+                repo['jobs_dir'] = "zuul.d"
+            elif os.path.isfile(os.path.join(repo.get('dest'),
+                                             ZUUL_JOBS_LOCATION[1])):
+                repo['jobs_file'] = ".zuul.yaml"
+
+    def get_file_list(self):
+        """
+        This method will give list of files in zuul jobs dir.
+        """
+        def get_yaml(x):
+            return [f for f in x
+                    if f.endswith('.yaml') or
+                    f.endswith(".yml")]
+        job_files = []
+        for repo in self.repos:
+            file_path = repo.get('jobs_file', None)
+            jobs_dir = repo.get('jobs_dir', None)
+            if file_path:
+                file_path = os.path.join(repo.get('dest'), file_path)
+            if jobs_dir:
+                jobs_dir = os.path.join(repo.get('dest'), jobs_dir)
+
+            if file_path and os.path.isfile(file_path):
+                job_files.append(file_path)
+
+            elif jobs_dir and os.path.isdir(jobs_dir):
+                files = os.listdir(jobs_dir)
+                job_files.extend([os.path.join(jobs_dir, i)
+                                  for i in get_yaml(files)])
+        return job_files
+
+    def _parse_file(self, file_path):
+        """
+        This method will parse single file
+        """
+        LOG.debug("Reading: {}".format(file_path))
+        if os.path.isfile(file_path):
+            try:
+                data = open(file_path, "rb").read()
+                return yaml.load(data, Loader=get_loader())
+            except Exception as e:
+                print(e)
+                LOG.error("Not able to read: {}".format(file_path))
+                # Printing only error message. We need to traverse all files
+                # Error message will tell which file not able to parse.
+        else:
+            raise FileNotFoundError("File not exist: {}".format(file_path))
+
+    def get_jobs(self):
+        """
+        This method will parse jobs from file
+        """
+        t = Tenant(name='openstack')
```

Comment:
Can we know the name of the tenant through the name of the repository? Or there is not always a repo per tenant to add the jobs?

Reply:
No. Tenant name is not defined in repos. I've used `openstack` as constant for now, may be we can change that in future.